libraries/chain/transaction_context.cpp

```diff
@@ -156,13 +156,13 @@ namespace eosio { namespace chain {
       if( initial_net_usage > 0 )
          add_net_usage( initial_net_usage );  // Fail early if current net usage is already greater than the calculated limit
 
-      checktime(); // Fail early if deadline has already been exceeded
-
       if(control.skip_trx_checks())
          deadline_timer.start(fc::time_point::maximum());
       else
          deadline_timer.start(_deadline);
 
+      checktime(); // Fail early if deadline has already been exceeded
+
```

Comment:
There is somewhat of a landmine above. The checktime timer needs to start before `checktime()` is ever called, otherwise the timer will still be still be in an expired state and cause the `checktime()` to fail. This wasn't a problem previously because an expired checktime timer simply caused `checktime()` to poll the time.

Not sure the best way of trying to guard against someone falling in to that trap in the future.

Reply:
Can the `deadline_timer` be something that the calling code constructs/initializes and moves into a `transaction_context` as part of construction? that would make it impossible to write a call to `transaction_context::checktime` without having a valid and initialized `deadline_timer`.  That would move the `deadline_timer.start` call quite a bit and I'm not sure what consequences that would have

Reply:
>Can the deadline_timer be something that the calling code constructs/initializes and moves into a transaction_context as part of construction?

Unless I'm misunderstanding, this won't work very well post #7449 because now checktime_timer on construction acquires a system resource i.e. it does some syscalls and thus has overhead I'd really like to avoid performing for each transaction. That is why the timer was moved to live in controller, so that it is only created once.

I could add something like a `checktime::reset()` that would set `expired = 0`; the transaction_context could call this in its ctor. This would ensure any `checktime()` calls before transaction_context got around to starting the timer would succeed. Still a bit hokey I know. Any other ideas?

Reply:
Right sorry, didn't remember its a passed in reference.  I think its the responsibility of the calling code (which is providing that reference) to make sure it provides a "ready to use" `checktime_timer`, I agree it's pretty fragile to implicitly assume someone has handled this for you so... my hair-brained idea is to create an abstraction around it. 

 Something like a `checktime_timer_pool` which creates the system resource(s) and "loans" them out to `checktime_timer` objects (which come into life properly reset).  Those objects would return the system resource to the pool when they are destroyed (if nothing stole its resource with `std::move`).  

`transaction_context` would then expect a `checktime_timer&&` in its constructor and move it to a member.  The calling frame would never get that object back so, it cannot pass it to another `transaction_context`.  It would always be calling `checktime_timer_pool::take_one()` or something.   This makes the lifetime of the `checktime_timer` "safe" from the point of view of the `transaction_context` and provides an RAII style lifecycle management of the expensive system resource and the expected `reset`

Reply:
I don't think we should get too carried away just yet with an actual pool. We all know multiple timers are going to be needed when multithreading comes but it's not clear at all how that truly comes together just yet. Don't want to try and predict the future too much here.

I tried to clean up the lack of RAII-ness by making controller move a transaction_checktime_timer in to the transaction_context upon its construction. This still isn't very pool-like since transaction_checktime_timer doesn't phone anything up to say it has been destroyed. But, at least it hides the initialization and cleanup from transaction_context unlike before where transaction_context had to manually create another scoped exit of sorts.